framework/Objects/RestController.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+namespace JustCoded\WP\Framework\Objects;
+
+
+/**
+ * Class RestController
+ * @package JustCoded\WP\Framework\Objects
+ *
+ * @method RestController instance() static
+ */
+abstract class RestController {
+	use Singleton;
+
+	/**
+	 * @var string
+	 */
+	protected $namespace = 'api';
+
+	/**
+	 * @var array
+	 */
+	protected $routes = [];
+
+	/**
+	 * @var array
+	 */
+	protected $slugs;
+
+	/**
+	 * Constructor
+	 * Check required parameter ROUTE and add WP action hook
+	 *
+	 * @throws \Exception Missing $ROUTE property.
+	 */
+	protected function __construct() {
+		if ( empty( $this->namespace ) ) {
+			throw new \Exception( 'REST ENDPOINT "' . static::class . '" init failed: missing namespace property' );
+		}
+
+		add_action( 'rest_api_init', [ $this, 'register_endpoint' ] );
+	}
+
+	/**
+	 * Get_permalink
+	 *
+	 * @param string $slug_key .
+	 * @param array  $args .
+	 *
+	 * @return string|bool
+	 */
+	public function get_permalink( $slug_key, $args = array() ) {
+		if ( is_array( $slug_key ) ) {
+			return false;
+		}
+
+		$data       = [];
+		$slug       = $this->slugs[ $slug_key ];
+		$slug_regex = str_replace( [ '(', ')' ], [ '_*(', ')*_' ], $slug );
+		$home_url   = home_url( '/wp-json/' . $this->namespace );
+
+		if ( ! preg_match_all( '/<(\w+)>/i', $slug, $args_matches ) ) {
+			return false;
+		}
+
+		if ( ! preg_match_all( '/_\*(.*?)\*_/i', $slug_regex, $data_matches ) ) {
+			return false;
+		}
+
+		foreach ( $args_matches[1] as $match ) {
+			if ( array_key_exists( $match, $args ) ) {
+				$data[ $match ] = $args[ $match ];
+				unset( $args[ $match ] );
+			}
+		}
+
+		$query_args = add_query_arg( [ $args ] );
+
+		foreach ( $data_matches[1] as $match ) {
+			if ( ! preg_match( '/<(\w+)>/i', $match, $key_match ) ) {
+				continue;
+			}
+
+			if ( ! empty( $data[ $key_match[1] ] ) ) {
+				$slug = str_replace( $match, $data[ $key_match[1] ], $slug );
+			}
+		}
+
+		return $home_url . $slug . $query_args;
```

Comment:
make all code here a lot easier. 

what you need:
1) replace WP route regular expression to a simple string like: /my-path/{my-var} (with a preg_replace)
2) just run foreach($args) and do str_replace.